src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -805,21 +805,21 @@ std::shared_ptr<ExpressionResult<Result>> ExpressionEvaluator::_evaluate_functio
     const FunctionExpression& expression) {
   switch (expression.function_type) {
     case FunctionType::Concatenate:
-    case FunctionType::Substring:
       if constexpr (std::is_same_v<Result, pmr_string>) {
-        if (expression.function_type == FunctionType::Substring) {
-          return _evaluate_substring(expression.arguments);
-        }
         return _evaluate_concatenate(expression.arguments);
       }
-
-      Fail("Function can only be evaluated to a string.");
+      Fail("CONCAT() can only be evaluated to a string.");
+    case FunctionType::Substring:
+      if constexpr (std::is_same_v<Result, pmr_string>) {
+        return _evaluate_substring(expression.arguments);
+      }
+      Fail("SUBSTR() can only be evaluated to a string.");
```

Comment:
"To a string"? Rather "on a string"?

Reply:
Also had that first. Technically, we determine the data type from the expression to be evalauted, not from the input expression(s). But I guess we can live with that small impreciseness in favor of a nicer error message. The expressions should already have complained if they were to be used on non-strings.